proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,138 @@
+# Feature name
+
+* Proposal: TBD
+* Author(s): [Kevin Ballard](https://github.com/kballard)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+Add 3 new `SequenceType` functions `scan(_:combine:)`, `takeWhile(_:)`, and
+`dropWhile(_:)`, with overrides as appropriate on `CollectionType`,
+`LazySequenceType`, and `LazyCollectionType`, as well as a global function
+`iterate(_:apply:)`.
+
+Swift-evolution thread:
+[Proposal: Add scan, takeWhile, dropWhile, and iterate to the stdlib](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20151228/004690.html)
+
+## Motivation
+
+The Swift standard library provides many useful sequence manipulators like
+`dropFirst(_:)`, `filter(_:)`, etc. but it's missing a few common methods that
+are quite useful.
+
+## Proposed solution
+
+Add the following extension to `SequenceType`:
+
+```swift
+extension SequenceType {
+  /// Returns an array containing the results of
+  ///
+  ///     p.reduce(initial, combine: combine)
+  ///
+  /// for each prefix `p` of `self`, in order from shortest to longest.
+  /// For example:
+  ///
+  ///     (1..<6).scan(0, combine: +) // [0, 1, 3, 6, 10, 15]
+  ///
+  /// - Complexity: O(N)
+  func scan<T>(initial: T, @noescape combine: (T, Self.Generator.Element) throws -> T) rethrows -> [T]
+}
+```
+
+Modify the declaration of `SequenceType` with two new members:
+
+```swift
+protocol SequenceType {
+  // ...
+  /// Returns a subsequence by skipping elements while `dropElement` returns
+  /// `true` and returning the remainder.
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  /// Returns a subsequence containing the elements until `takeElement` returns
+  /// `false` and skipping the remainder.
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
```

Comment:
I don't like inverting the predicate here. The focus here is on retaining the initial elements according to some predicate, and it feels very backwards to flip the predicate around so it describes what you aren't retaining. In my experience, usually the predicate with this function is used to describe a characteristic of the elements you want to keep, not a characteristic of the elements you want to skip.

I could be convinced to go with `prefixHaving(_:)` (or is it `prefix(having:)` now?), though I'm still partial to `takeWhile(_:)` because that's the naming that all of the precedent I'm familiar with uses.